src/ca/notifierConveyor.cpp

```diff
@@ -0,0 +1,94 @@
+/**
+ * Copyright - See the COPYRIGHT that is included with this distribution.
+ * pvAccessCPP is distributed subject to a Software License Agreement found
+ * in file LICENSE that is included with this distribution.
+ */
+
+#include <iostream>
+#include <queue>
+#include <epicsThread.h>
+#include <pv/event.h>
+#include <pv/lock.h>
+
+#define epicsExportSharedSymbols
+#include "notifierConveyor.h"
+
+using epics::pvData::Lock;
+
+namespace epics {
+namespace pvAccess {
+namespace ca {
+
+NotifierConveyor::~NotifierConveyor()
+{
+    if (thread) {
+        {
+            Lock the(mutex);
+            halt = true;
+        }
+        workToDo.signal();
+        thread->exitWait();
+    }
+}
+
+void NotifierConveyor::start()
+{
+    if (thread) return;
+    thread = std::tr1::shared_ptr<epicsThread>(new epicsThread(*this,
+        "caProvider::clientNotifier",
+        epicsThreadGetStackSize(epicsThreadStackBig),
+        epicsThreadPriorityLow));
+    thread->start();
+}
+
+void NotifierConveyor::notifyClient(
+    NotificationPtr const &notificationPtr)
+{
+    {
+        Lock the(mutex);
+        if (halt || notificationPtr->queued) return;
+        notificationPtr->queued = true;
+        workQueue.push(notificationPtr);
+    }
+    workToDo.signal();
+}
+
+void NotifierConveyor::run()
+{
+    bool stopping;
+    do {
+        workToDo.wait();
+        Lock the(mutex);
+        stopping = halt;
+        bool work = !workQueue.empty();
+        while (work)
+        {
+            NotificationWPtr notificationWPtr(workQueue.front());
+            workQueue.pop();
+            work = !workQueue.empty();
+            NotificationPtr notification(notificationWPtr.lock());
+            if (notification) {
+                notification->queued = false;
+                NotifierClientPtr client(notification->client.lock());
+                if (client) {
+                    the.unlock();
```

Comment:
Think about client's destructor trying to un-queue itself, lock held or not?

Reply:
There is no API for a client to unqueue itself, only the conveyor thread can remove items from the queue.

Looking at `NotifierConveyor::run()` the mutex is taken at the top of the outer loop for access to the conveyor's data members, and unlocked before calling a `NotifierClient::notifyClient()` method. After that returns if there were more items on the queue (having read that into `work` before we unlocked) we must take the lock again before we loop back to get the next item. If the queue was empty we will fall out of the inner loop without taking the lock again. Thus when the `NotifierClientPtr client` goes out of scope we have two cases to consider, since the lock may be held or not depending whether `work` was set.

If `client` holds the last reference to a `NotifierClient` object the `shared_ptr` destructor might result in other objects being deleted, the interesting ones here being `Notification` and/or `NotifierConveyor`.

Our strong reference `notification` will slightly out-last the `client` object, but it's destruction has no knock-on effects.

The `CAChannelProvider` contains the two `NotifierConveyor` data members. Are there any restrictions on deleting a ChannelProvider from one of its Requester callbacks?

If this `NotifierConveyor` object gets deleted by the `NotifierClient` destructor, the conveyor's destructor will attempt to take the mutex. It doesn't matter whether we already own it or not though because the recursive nature of the mutex permits the destructor to acquire the lock immediately or after a short delay if some other thread is queueing a notification. The destructor sets `halt`, releases the lock and signals the Event, then calls `thread->exitWait()`. The semantics of that `epicsThread` method ensure that since it is being called from inside the thread it will return immediately, permitting the destructor to exit. At this point the conveyor object data members may no longer be accessed, which is a problem if `work` is true or `stopping` hasn't been set from `halt` by then.

Changing the `run()` method to solve that...

Reply:
I think I may not have made myself clear.  Of course you can, with sufficient effort, convince yourself that a particular scenario is not a concern.

The point I was trying to make, and which may have been lost in the weeds, is that proper attention to _local_ shared_ptr scope can avoid the need to perform a _global_ analysis.  This is a matter of better practice (I won't claim it's best) which I've come to after wrestling with this code base for the past ~3 years.  As with locking (I guess I failed to draw that analogy) I find it easier to work with rules which I can apply to pieces of code in isolation.

To return to the specific, I would write the skeleton of a worker thread of this sort as:

```c++
struct myworker {
    epicsMutex lock;
    epicsEvent wakeup;
    std::list<std::weak_ptr<...>> work;
    bool running;

    void run() {
        epicsGuard<epicsMutex> G(lock);
        while(running) {
            {
                epicsGuardRelease<epicsMutex> U(G);
                wakeup.wait();
            }
            while(!work.empty()) {
                std::weak_ptr<...> wjob(work.front);
                work.pop_front();
                {
                    epicsGuardRelease<epicsMutex> U(G);
                    std::shared_ptr<...> job(wjob.lock());
                    try { if(job) job->whatever(); } catch ...
                }
            }
        }
    }
```

The key points being that the lock is held in `run()` by default, and only released when specifically needed.  Also, strong references to the work to be executed only exist in the same scope as the `epicsGuardRelease`.  I put this outside of the innermost try/catch since I often want to use some aspect in a log message.